### PR TITLE
 Resolve race between OnReadDone(ok=false) and IsCancelled

### PR DESCRIPTION
**Pull request by vjpai**
Because the OnReadDone notification and the OnCancel notification come in separate batches from core, there can be a condition where an OnReadDone returns ok=false because the stream was cancelled but the server can't tell if it was because the stream was cancelled or actually closed cleanly. This PR resolves this issue by specifically marking the places in the transport where OnReadDone will return an empty message because of a stream failure and providing a core subsurface function to get this information back out of the grpc_call struct. Because it is subsurface, this does not require a gRFC. Of course, it can also only be used at the current time for C++ but that is the only place where we have a customer feature request for this information anyway.

Originally identified by @mjs476 as b/138186533

### CODE REVIEW

**Review by ctiller — APPROVED** [review]

**Comment by ctiller** [comment]
LGTM, but let's wait for @markdroth before submitting because I'm likely still missing context.

**Comment by vjpai** [comment]
@ctiller Let me add that this resolves b/138186533

File: test/cpp/microbenchmarks/bm_chttp2_transport.cc [hunk 1]
```diff
@@ -565,6 +565,7 @@ static void BM_TransportStreamRecv(benchmark::State& state) {
   grpc_transport_stream_op_batch_payload op_payload(nullptr);
   grpc_transport_stream_op_batch op;
   grpc_core::OrphanablePtr<grpc_core::ByteStream> recv_stream;
+  bool recv_error = false;
```
Comment (markdroth) [inline test/cpp/microbenchmarks/bm_chttp2_transport.cc:568]: If the transports check for this being null, then there won't be any need to set this here.

**Review by markdroth — COMMENTED** [review]
This approach looks good!  Comments are mostly about details more than overall structure.

Please let me know if you have any questions.  Thanks!

File: src/cpp/server/server_context.cc [hunk 2]
```diff
@@ -324,10 +325,17 @@ void ServerContextBase::TryCancel() const {
   }
 }
 
+void ServerContextBase::MaybeMarkCancelledOnRead() {
+  if (grpc_call_recv_message_error(call_.call)) {
+    marked_cancelled_.store(true, std::memory_order_release);
```
[the remark this answers is not in the document]
Comment (soheilhy) [reply src/cpp/server/server_context.cc:330]: Thanks for including! I checked the code and AFAICT the use of the atomic here is safe.

File: src/core/ext/transport/chttp2/transport/chttp2_transport.cc [hunk 3]
```diff
@@ -1947,6 +1948,8 @@ void grpc_chttp2_maybe_complete_recv_message(grpc_chttp2_transport* /*t*/,
       null_then_sched_closure(&s->recv_message_ready);
     } else if (s->published_metadata[1] != GRPC_METADATA_NOT_PUBLISHED) {
       *s->recv_message = nullptr;
+      *s->recv_message_error =
```
[the remark this answers is not in the document]
Comment (vjpai) [reply src/core/ext/transport/chttp2/transport/chttp2_transport.cc:1951]: Great observation, and that takes care of the retry and health check comments as well.

File: src/cpp/server/server_context.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (vjpai) [reply src/cpp/server/server_context.cc:330]: Thanks for checking this out @soheilhy!

File: include/grpcpp/impl/codegen/server_context.h [hunk 4]
```diff
@@ -525,6 +527,8 @@ class ServerContextBase {
       default_reactor_;
   std::atomic_bool default_reactor_used_{false};
   std::unique_ptr<TestServerCallbackUnary> test_unary_;
+
+  std::atomic_bool marked_cancelled_{false};
```
[the remark this answers is not in the document]
Comment (vjpai) [reply include/grpcpp/impl/codegen/server_context.h:531]: Cool!

File: src/core/ext/filters/client_channel/health/health_check_client.h [hunk 5]
```diff
@@ -124,6 +124,7 @@ class HealthCheckClient : public InternallyRefCounted<HealthCheckClient> {
 
     // recv_message
     OrphanablePtr<ByteStream> recv_message_;
+    bool recv_message_error_;
     grpc_closure recv_message_ready_;
     grpc_slice_buffer recv_message_buffer_;
     Atomic<bool> seen_response_{false};
```
[the remark this answers is not in the document]
Comment (vjpai) [reply src/core/ext/filters/client_channel/health/health_check_client.h:130]: recv_message_error_ is now removed, but I did this for cancelled_

File: src/core/lib/transport/transport.h [hunk 6]
```diff
@@ -295,6 +295,8 @@ struct grpc_transport_stream_op_batch_payload {
     // containing a received message.
     // Will be NULL if trailing metadata is received instead of a message.
     grpc_core::OrphanablePtr<grpc_core::ByteStream>* recv_message = nullptr;
+    // Was this recv_message failed for reasons other than a clean end-of-stream
+    bool* recv_message_error = nullptr;
```
[the remark this answers is not in the document]
Comment (vjpai) [reply src/core/lib/transport/transport.h:299]: Thanks so much for this comment and the others like it. I think this makes it a much more clear reading.

File: src/core/ext/filters/client_channel/retry_filter.cc [hunk 7]
```diff
@@ -1650,6 +1653,8 @@ void RetryFilter::CallData::CallAttempt::BatchData::
   ++call_attempt_->started_recv_message_count_;
   batch_.recv_message = true;
   batch_.payload->recv_message.recv_message = &call_attempt_->recv_message_;
+  batch_.payload->recv_message.recv_message_error =
```
[the remark this answers is not in the document]
Comment (vjpai) [reply src/core/ext/filters/client_channel/retry_filter.cc:1656]: I've removed these client-only uses of this code and check for nullptr in the transport.

**Comment by vjpai** [comment]
Thanks for the reviews, @markdroth and @soheilhy .

@yashykt can I get a review on the chttp2 transport code's use of  `GRPC_METADATA_PUBLISHED_AT_CLOSE` from Mark's comment above.

**Comment by vjpai** [comment]
I'm also going to ask for these topics not to get migrated to the core surface API for now on the YAGNI principle and for expediency in timing.

**Comment by markdroth** [comment]
> I'm also going to ask for these topics not to get migrated to the core surface API for now on the YAGNI principle and for expediency in timing.

Okay; this isn't optimal, but we can probably live with it as a short-term thing.  But can we at least add a TODO somewhere about exposing this in the C-core surface API if it winds up being needed for wrapped languages?  I think this could come up whenever someone wants to write a gRPC proxy in a wrapped language, depending on the API of the wrapped language.

**Review by markdroth — APPROVED** [review]
Thanks for doing this!

**Comment by vjpai** [comment]
Thanks for the reviews, @markdroth @soheilhy and @ctiller . I'll wait to merge until I get your review @yashykt

**Review by yashykt — APPROVED** [review]

**Comment by vjpai** [comment]
Thanks!

**Comment by sb230132** [comment]
Can there be a race between Finish() and OnCancel() ?

**Pull request merged**